# Meridional Heat Transport (MHT)
This notebook calculates the model MHT from two ways, based on distinct MOM5 diagnostics: `temp_yflux_adv_int_z` (depth-integrated meridional heat transport due to resovled advection) and `net_sfc_heating` (net surface heat flux). To use this notebook, we need to ensure this diagnostic is available in the model output. We can check if a variable is available using `cc.querying.get_variables()` function from the COSIMA cookbook.  
  
Note that currently this notebook calculates the total (all basins) MHT and it also includes comparisons to a few observational products.

## Loading relevant libraries

In [1]:
#Accessing the COSIMA database
import cosima_cookbook as cc

#Manipulating and analysing data
import numpy as np
import pandas as pd
import netCDF4 as nc
import xarray as xr

#Regular expressions (regex) to extract information from strings
import re

#Plotting figures
import matplotlib.pyplot as plt
#Shows graph in notebook
%matplotlib inline

## Loading model data

We will use the COSIMA cookbook to search its database and load `temp_yflux_adv_int_z` to start our calculations. To do this we will need to do the following:
1. Start a COSIMA cookbook session
2. Define the model configuration of interest
3. Define the experiment of interest
4. Define the diagnostic (variable) of interest

**NOTE:** If you are in doubt about the models, experiments and diagnostics available in the database, check the [Cookbook Tutorial]("../Tutorials/COSIMA_CookBook_Tutorial.ipynb") for more information.

In [2]:
#Start a COSIMA cookbook session
session = cc.database.create_session()

#Define experiment of interest
experiment = '025deg_jra55v13_iaf_gmredi6'

We are now ready to query the database and load the data to start our analysis. We load `temp_yflux_adv_int_z`. For this example, we have chosen to use 6 years of output.

In [3]:
mht = cc.querying.getvar(experiment, 'temp_yflux_adv_int_z', session, n=3)
mht

<xarray.DataArray 'temp_yflux_adv_int_z' (time: 72, yu_ocean: 1080,
                                          xt_ocean: 1440)>
dask.array<concatenate, shape=(72, 1080, 1440), dtype=float32, chunksize=(1, 540, 720), chunktype=numpy.ndarray>
Coordinates:
  * yu_ocean  (yu_ocean) float64 -81.02 -80.92 -80.81 ... 89.79 89.89 90.0
  * time      (time) datetime64[ns] 2198-01-14T12:00:00 ... 2203-12-14T12:00:00
  * xt_ocean  (xt_ocean) float64 -279.9 -279.6 -279.4 ... 79.38 79.62 79.88
Attributes:
    long_name:      z-integral of cp*rho*dxt*v*temp
    units:          Watts
    valid_range:    [-1.e+18  1.e+18]
    cell_methods:   time: mean
    time_avg_info:  average_T1,average_T2,average_DT
    coordinates:    geolon_t geolat_c
    ncfiles:        ['/g/data/hh5/tmp/cosima/access-om2-025/025deg_jra55v13_i...

Viewing the variable containing our data provides a number of key information, including:
- *Shape of dataset* - In this case the data includes 72 time steps, 1080 steps along the y axis and 1440 along the x axis.
- *Name of coordinates* - Our data includes `time`, `yu_ocean` (i.e., latitude), and `xt_ocean` (i.e., longitude).
- *Values included under each coordinate* - We can see that `time` includes monthly values from 2198-01-14 to 2203-12-14.
- *Metadata* - This information is included under `Attributes` and it includes things like units.

## Calculating mean and converting units
From the metadata, we can see that our dataset is in Watts (W), so we will convert them to petawatts (PW)

In [4]:
#We transform units from watts to petawatts
mht = mht * 1e-15
mht = mht.assign_attrs(units='PettaWatts')
mht

<xarray.DataArray 'temp_yflux_adv_int_z' (time: 72, yu_ocean: 1080,
                                          xt_ocean: 1440)>
dask.array<mul, shape=(72, 1080, 1440), dtype=float32, chunksize=(1, 540, 720), chunktype=numpy.ndarray>
Coordinates:
  * yu_ocean  (yu_ocean) float64 -81.02 -80.92 -80.81 ... 89.79 89.89 90.0
  * time      (time) datetime64[ns] 2198-01-14T12:00:00 ... 2203-12-14T12:00:00
  * xt_ocean  (xt_ocean) float64 -279.9 -279.6 -279.4 ... 79.38 79.62 79.88
Attributes:
    units:    PettaWatts

and then we calculate the mean across the `time` dimension.

In [5]:
#We calculate the mean across all timesteps
mht = mht.mean('time')

#Finally, we sum up all values across each longitude
mht_xsum = mht.sum('xt_ocean')
mht_xsum

<xarray.DataArray 'temp_yflux_adv_int_z' (yu_ocean: 1080)>
dask.array<sum-aggregate, shape=(1080,), dtype=float32, chunksize=(540,), chunktype=numpy.ndarray>
Coordinates:
  * yu_ocean  (yu_ocean) float64 -81.02 -80.92 -80.81 ... 89.79 89.89 90.0

# Estimating meridional heat transport from the net surface heat flux (assuming steady state)

First, we load the surface heat flux and grid metrics:

In [6]:
shflux = cc.querying.getvar(experiment, 'net_sfc_heating', session, n=3)
shflux_am = shflux.mean('time').load()
area = cc.querying.getvar(experiment, 'area_t', session, ncfile="ocean_grid.nc", n=1)
lat = cc.querying.getvar(experiment, 'geolat_t', session, ncfile="ocean_grid.nc", n=1)
latv = cc.querying.getvar(experiment, 'yt_ocean', session, ncfile="ocean_grid.nc", n=1)



Now calculate Meridional Heat Flux (MHF):

$\textrm{MHF} = \textrm{Cummulative sum of } (\textrm{SHFLUX} \times \textrm{AREA}) \textrm{ along latitudes}$

**Note**: The following cell might take 1-2 min.

In [ ]:
%time

mhf = xr.zeros_like(latv)

for i in range(len(latv)):
    inds = lat < latv[i]
    atmp = area.where(lat < latv[i])
    stmp = shflux_am.where(lat < latv[i])
    mhf[i] = np.sum(atmp * stmp)

mhf = mhf + (mhf[0] - mhf[-1])/2


CPU times: user 4 µs, sys: 3 µs, total: 7 µs
Wall time: 18.4 µs


In [ ]:
#We transform units from watts to petawatts
mhf = mhf * 1e-15
mhf = mhf.assign_attrs(units='PettaWatts')
mhf

## Comparison between model output and observations
Before producing our figure, we will compare the model output with observations to check the model accuracy.

### Read ERBE Period Ocean and Atmospheric Heat Transport
Observations are also available in Gadi, here we show how to load them to our notebook.

In [ ]:
#Path to the file containing observations
filename = '/g/data3/hh5/tmp/cosima/observations/original/MHT/obs_vq_am_estimates.txt'

#Creating empty variables to store our observations
erbe_mht = []
erbe_lat = []

#Opening data and saving it to empty variables above
with open(filename) as f:
    #Open each line from rows 1 to 96
    for line in f.readlines()[1:96]:
        #Separating each line to extract data
        line = line.strip()
        sline = line.split()
        #Extracting latitude and MHT and saving to empty variables
        erbe_lat.append(float(sline[0]))
        erbe_mht.append(float(sline[3]))

#Checking MHT variables
erbe_mht

### Read NCEP and ECMWF Oceanic and Atmospheric Transport Products

In [ ]:
#Path to the file containing observations
filename = '/g/data3/hh5/tmp/cosima/observations/original/MHT/ANNUAL_TRANSPORTS_1985_1989.ascii'

#Creating empty variables to store our observations
ncep_g_mht = []
ecwmf_g_mht = []
ncep_g_err = []
ecwmf_g_err = []
ncep_a_mht = []
ecwmf_a_mht = []
ncep_a_err = []
ecwmf_a_err = []
ncep_p_mht = []
ecwmf_p_mht = []
ncep_p_err = []
ecwmf_p_err = []
ncep_i_mht = []
ecwmf_i_mht = []
ncep_i_err = []
ecwmf_i_err = []
ncep_ip_mht = []
ecwmf_ip_mht = []
ncep_ip_err = []
ecwmf_ip_err = []
o_lat = []

#Opening data and saving it to empty variables above
with open(filename) as f:
#Open each line in file (ignoring the first row)
    for line in f.readlines()[1:]:
        #Separating each line to extract data
        line = line.strip()
        sline = line.split()
        #Extracting values and saving to correct variable defined above
        o_lat.append(float(sline[0])*0.01)  # T42 latitudes (north to south)
        ncep_g_mht.append(float(sline[4])*0.01) # Residual Ocean Transport - NCEP
        ecwmf_g_mht.append(float(sline[5])*0.01)# Residual Ocean Transport - ECWMF
        ncep_a_mht.append(float(sline[7])*0.01)# Atlantic Ocean Basin Transport - NCEP
        ncep_p_mht.append(float(sline[8])*0.01)# Pacific Ocean Basin Transport - NCEP
        ncep_i_mht.append(float(sline[9])*0.01)# Indian Ocean Basin Transport - NCEP
        ncep_g_err.append(float(sline[10])*0.01)# Error Bars for NCEP Total Transports
        ncep_a_err.append(float(sline[11])*0.01)# Error Bars for NCEP Atlantic Transports 
        ncep_p_err.append(float(sline[12])*0.01)# Error Bars for NCEP Pacific Transports 
        ncep_i_err.append(float(sline[13])*0.01)# Error Bars for NCEP Indian Transports 
        ecwmf_a_mht.append(float(sline[15])*0.01)# Atlantic Ocean Basin Transport - ECWMF
        ecwmf_p_mht.append(float(sline[16])*0.01)# Pacific Ocean Basin Transport - ECWMF
        ecwmf_i_mht.append(float(sline[17])*0.01)# Indian Ocean Basin Transport - ECWMF
        ecwmf_g_err.append(float(sline[18])*0.01)# Error Bars for ECWMF Total Transports
        ecwmf_a_err.append(float(sline[19])*0.01)# Error Bars for NCEP Atlantic Transports
        ecwmf_p_err.append(float(sline[20])*0.01)# Error Bars for NCEP Pacific Transports
        ecwmf_i_err.append(float(sline[21])*0.01)# Error Bars for NCEP Indian Transports

#Calculating MHT
ncep_ip_mht = [a+b for a, b in zip(ncep_p_mht,ncep_i_mht)]
ecwmf_ip_mht = [a+b for a, b in zip(ecwmf_p_mht,ecwmf_i_mht)]
ncep_ip_err = [max(a,b) for a, b in zip(ncep_p_err,ncep_i_err)]
ecwmf_ip_err = [max(a,b) for a, b in zip(ecwmf_p_err,ecwmf_i_err)]

## Plotting model outputs against observations

We plot the global meridional heat transport as calculated from model outputs (blue line) and observations.

In [ ]:
fig = plt.figure(figsize=(12, 6))
ax = fig.add_subplot(1, 1, 1)

#Plotting model outputs
mht_xsum.plot(ax = ax, color="blue", label="ACCESS-OM2-025")
ax.plot(latv,mhf, 'g', label="ACCESS-OM2-025 (shflux)")

#Adding observations and error bars for observations
ax.plot(erbe_lat, erbe_mht, 'k--', linewidth=1, label="ERBE, JRA-25, NCEP/NCAR and ERA40")
plt.errorbar(o_lat[::-1], ncep_g_mht[::-1], yerr=ncep_g_err[::-1], c='gray', fmt='s', 
             markerfacecolor='k', markersize=3, capsize=2, linewidth=1, label="NCEP")
plt.errorbar(o_lat[::-1], ecwmf_g_mht[::-1], yerr=ecwmf_g_err[::-1], c='gray', fmt='D', 
             markerfacecolor='white', markersize=3, capsize=2, linewidth=1, label="ECWMF")
plt.errorbar(24, 1.5, yerr=0.3, fmt='o', c='black', markersize=3, capsize=2, linewidth=1, 
             label="Macdonald and Wunsch 1996")
plt.errorbar(-30, -0.9, yerr=0.3, fmt='o', c='black', markersize=3, capsize=2, linewidth=1)
plt.errorbar(24, 2.0, yerr=0.3, fmt='x', c='green', markersize=3, capsize=2, linewidth=1, 
             label="Lavin et al. and Bryden et al.")
plt.errorbar(24, 1.83, yerr=0.28, fmt='^', c='red', markersize=4, capsize=2, linewidth=1, 
             label="Ganachaud and Wunsch 2003")
plt.errorbar(-30, -0.6, yerr=0.3, fmt='^', c='red', markersize=4, capsize=2, linewidth=1)
plt.errorbar(-19, -0.8, yerr=0.3, fmt='^', c='red', markersize=4, capsize=2, linewidth=1)
plt.errorbar(47, 0.6, yerr=0.1, fmt='^', c='red', markersize=4, capsize=2, linewidth=1)

#Adding legend
plt.legend(frameon=False, fontsize=12)
plt.axhline(y=0, linewidth=1, color='black')

#Defining plot limits along the y axis
plt.ylim(-2.25, 2.5)

#Adding titles for figure and axes
plt.title('Global Ocean Meridional Heat Transport', fontsize=18)
plt.xlabel('Latitude')
plt.ylabel('$10^{15}$ Watts');